tests/phpunit/tests/admin/wpUpgrader.php

```diff
@@ -0,0 +1,1430 @@
+									),
+									'subsubfile4.php' => array(
+										'name' => 'subsubfile4.php',
+									),
+								),
+							),
+						),
+					),
+				),
+				'path'         => 'custom_path/',
+			),
+		);
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::clear_destination()` returns early with `true`
+	 * when the destination does not exist.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::clear_destination
+	 */
+	public function test_clear_destination_should_return_early_when_the_destination_does_not_exist() {
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'is_writable' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'chmod' );
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'delete' );
+
+		$destination = DIR_TESTDATA . '/upgrade/';
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'dirlist' )
+				->with( $destination )
+				->willReturn( false );
+
+		$this->assertTrue( $this->instance->clear_destination( $destination ) );
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::clear_destination()` clears
+	 * the destination directory.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::clear_destination
+	 */
+	public function test_should_clear_the_destination_directory() {
+		$destination = DIR_TESTDATA . '/upgrade/';
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'dirlist' )
+				->with( $destination )
+				->willReturn( array() );
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'delete' )
+				->with( $destination )
+				->willReturn( true );
+
+		$this->assertTrue( $this->instance->clear_destination( $destination ) );
+	}
+
+	/**
+	 * Tests that `WP_Upgrader::clear_destination()` returns a WP_Error object
+	 * if files are not writable.
+	 *
+	 * @ticket 54245
+	 *
+	 * @covers WP_Upgrader::clear_destination
+	 */
+	public function test_should_return_wp_error_if_files_are_not_writable() {
+		$this->instance->generic_strings();
+
+		$this->wp_filesystem_mock->expects( $this->never() )->method( 'delete' );
+
+		if ( ! defined( 'FS_CHMOD_DIR' ) ) {
+			define( 'FS_CHMOD_DIR', 0755 );
+		}
+
+		if ( ! defined( 'FS_CHMOD_FILE' ) ) {
+			define( 'FS_CHMOD_FILE', 0644 );
+		}
+
+		$destination = DIR_TESTDATA . '/upgrade/';
+		$dirlist     = array(
+			'file1.php' => array(
+				'name' => 'file1.php',
+				'type' => 'f',
+			),
+			'subdir'    => array(
+				'name' => 'subdir',
+				'type' => 'd',
+			),
+		);
+
+		$this->wp_filesystem_mock
+				->expects( $this->once() )
+				->method( 'dirlist' )
+				->with( $destination )
+				->willReturn( $dirlist );
+
+		$unwritable_checks = array(
+			array( $destination . 'file1.php' ),
+			array( $destination . 'file1.php' ),
+			array( $destination . 'subdir' ),
+			array( $destination . 'subdir' ),
+		);
+
+		$this->wp_filesystem_mock
+				->expects( $this->exactly( 4 ) )
+				->method( 'is_writable' )
+				->withConsecutive( ...$unwritable_checks )
```

Comment:
Be aware that `withConsecutive()` was deprecated in PHPUnit 9 and is being removed in PHPUnit 10.
Refs:
* https://github.com/sebastianbergmann/phpunit/issues/4255#issuecomment-636422439
* https://github.com/sebastianbergmann/phpunit/issues/4564

Reply:
Thanks for that info! I'll have to see what's a suitable replacement for this then.

Reply:
Update: I'm yet to find a suitable replacement for this. From what I can see in the references above, an alternative hasn't been implemented, but I may be missing something. Not sure of the path to move this part forward.

Reply:
@jrfnl As the WP test suite doesn't yet support PHPUnit 10 (via it's use of Yoast/PHPUnit-Polyfills 1.1), do you think it's worth putting this is as is to get some tests in for the class?

It doesn't look like there's a replacement for `withConsecutive` yet but I see you did some work on PHPCSStandards/PHPCSUtils to manage that so let me know if you have other suggestions.

Reply:
@peterwilsoncc For now, I'd just use the `withConsecutive()`.

If/when we update the test suite for PHPUnit 10.x with Polyfills 2.x, we can implement a solution for all the places where `withConsecutive()` is being used in one go. That should also make sure that we don't get fragmented solutions all over the place with people already trying to avoid `withConsecutive()`, but there not being a replacement.

As a side-note: the reason for not updating the WP Core test suite for PHPUnit 10.x so far, is the largely related to the changes in PHPUnit 10.x which make it neigh impossible to reliable check for PHP deprecations/notices/warnings in PHPUnit 10.x. That, combined with PHPUnit 9.x still running fine on PHP 8.3, means we don't have a reason at this time to update the test suite.

The PHP deprecations/notices/warnings change in PHPUnit 10.x means we won't be able to fail the test suite on those anymore (= problematic for WP) or if we turn the option on to fail the test suite, the tests suite would also fail on PHPUnit native warnings, which is something we really don't want.
On top of that we actually have tests setting expectations for PHP deprecations/notices/warning and those will also become problematic.

All in all, enough reason not to update (yet) if we don't have to.

Reply:
> @peterwilsoncc For now, I'd just use the `withConsecutive()`.
> 
> If/when we update the test suite for PHPUnit 10.x with Polyfills 2.x, we can implement a solution for all the places where `withConsecutive()` is being used in one go. That should also make sure that we don't get fragmented solutions all over the place with people already trying to avoid `withConsecutive()`, but there not being a replacement.

Thanks @jrfnl, I think I'll merge in trunk and commit what's in the PR to trunk and the 6.4 branch provided everything still passes, that will avoid what we  have slipping another release.